From f561644d9932dfe19dc3e64e4ec257d8b14c707b Mon Sep 17 00:00:00 2001
From: me-cafebabe <me.cafebabe@gmail.com>
Date: Thu, 5 Oct 2023 19:38:27 +0000
Subject: [PATCH] libc: libstdc++: Introduce libstdc++_vendor

A lot of blobs still link to libstdc++.so even after 8.1, so allow
devices to build a vendor copy of it.

Adding vendor_available to libstdc++ is not an option after Android 14,
since libstdc++ then gets listed on /vendor/etc/linker.config.pb,
which is duplicated with /system/etc/linker.config.pb and as a result
linkerconfig fails and stops the entire system from booting.

[    5.840677] DEBUG: Abort message: 'duplicate: libstdc++.so is provided by default and system in [vendor]'

Change-Id: Ib6eee48de87100c2fb0dd765d57ee95d3159c4ef
---
 libc/Android.bp | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/libc/Android.bp b/libc/Android.bp
index f5624efc1..5bca48482 100644
--- a/libc/Android.bp
+++ b/libc/Android.bp
@@ -1851,7 +1851,7 @@ cc_library_headers {
 // libstdc++.so and libstdc++.a.
 // ========================================================
 
-cc_library {
+cc_defaults {
     defaults: [
         "libc_defaults",
         "bug_24465209_workaround",
@@ -1862,7 +1862,7 @@ cc_library {
         "bionic/__cxa_pure_virtual.cpp",
         "bionic/new.cpp",
     ],
-    name: "libstdc++",
+    name: "libstdc++_defaults",
     static_ndk_lib: true,
     static_libs: ["libasync_safe"],
     apex_available: [
@@ -1899,6 +1899,17 @@ cc_library {
     },
 }
 
+cc_library {
+    name: "libstdc++",
+    defaults: ["libstdc++_defaults"],
+}
+
+cc_library_shared {
+    name: "libstdc++_vendor",
+    defaults: ["libstdc++_defaults"],
+    soc_specific: true,
+}
+
 genrule {
     name: "libstdc++.arm.map",
     out: ["libstdc++.arm.map.txt"],
-- 
2.39.5

